include/boost/geometry/algorithms/detail/buffer/buffer_policies.hpp

```diff
@@ -157,54 +151,116 @@ struct buffer_turn_info
     bool within_original;
     signed_size_type count_in_original; // increased by +1 for in ext.ring, -1 for int.ring
 
-    std::size_t count_on_offsetted;
-    std::size_t count_on_helper;
-    std::size_t count_within_near_offsetted;
-
     inline buffer_turn_info()
         : turn_index(0)
-        , location(location_ok)
-        , count_within(0)
+        , is_turn_traversable(true)
+        , close_to_offset(false)
+        , is_linear_end_point(false)
         , within_original(false)
         , count_in_original(0)
-        , count_on_offsetted(0)
-        , count_on_helper(0)
-        , count_within_near_offsetted(0)
     {}
 };
 
-struct buffer_operation_less
+struct buffer_less
+{
+    template <typename Indexed>
+    inline bool operator()(Indexed const& left, Indexed const& right) const
+    {
+        if (! (left.subject->seg_id == right.subject->seg_id))
+        {
+            return left.subject->seg_id < right.subject->seg_id;
+        }
+
+        // Both left and right are located on the SAME segment.
+        if (! (left.subject->fraction == right.subject->fraction))
+        {
+            return left.subject->fraction < right.subject->fraction;
+        }
+
+        return left.turn_index < right.turn_index;
+    }
+};
+
+struct piece_get_box
+{
+    template <typename Box, typename Piece>
+    static inline void apply(Box& total, Piece const& piece)
+    {
+        if (piece.m_piece_border.m_has_envelope)
+        {
+            geometry::expand(total, piece.m_piece_border.m_envelope);
+        }
+    }
+};
+
+template <typename DisjointBoxBoxStrategy>
+struct piece_ovelaps_box
 {
-    template <typename Turn>
-    inline bool operator()(Turn const& left, Turn const& right) const
+    template <typename Box, typename Piece>
+    static inline bool apply(Box const& box, Piece const& piece)
     {
-        segment_identifier const& sl = left.seg_id;
-        segment_identifier const& sr = right.seg_id;
+        if (piece.type == strategy::buffer::buffered_flat_end
+            || piece.type == strategy::buffer::buffered_concave)
+        {
+            // Turns cannot be inside a flat end (though they can be on border)
+            // Neither we need to check if they are inside concave helper pieces
+
+            // Skip all pieces not used as soon as possible
+            return false;
+        }
+        if (! piece.m_piece_border.m_has_envelope)
+        {
+            return false;
+        }
+
+        return ! geometry::detail::disjoint::disjoint_box_box(box, piece.m_piece_border.m_envelope,
+                                                              DisjointBoxBoxStrategy());
+    }
+};
 
-        // Sort them descending
-        return sl == sr
-            ? left.fraction < right.fraction
-            : sl < sr;
+struct turn_get_box
+{
+    template <typename Box, typename Turn>
+    static inline void apply(Box& total, Turn const& turn)
+    {
+        geometry::expand(total, turn.point);
```

Comment:
ExpandBoxPoint strategy is needed here.

Reply:
Not supported... There is something needed in the handling of expand with a strategy and a point. I leave this to you to avoid expanding the scope of this PR

Reply:
The expand strategies do not have the Earth model anyway (is there ever a reason to do that?).
So I can define and use them locally here, they don't need to be passed from the IntersectionStrategy.

Reply:
> The expand strategies do not have the Earth model anyway

That's because they are the same as the spherical ones and the unit sphere is enough as a model. Still it's different than cartesian.

Reply:
> 
> 
> > The expand strategies do not have the Earth model anyway
> 
> That's because they are the same as the spherical ones and the unit sphere is enough as a model. Still it's different than cartesian.

Yep - so I think what I now submitted (and merged) is fine?

Reply:
Almost ;) The CS should be taken from the strategy passed by the user instead of a geometry.

Reply:
See my message in the other PR - the passed strategy doesn't compile for points, because it's defined as a segment-specific strategy (for boxes it does compile but still wrong because a box is not a segment)

Reply:
I'm ok with leaving it as it is. I'll go back into it later if you don't want to deal with it. But I'll describe what we're dealing with here for informative purposes.

I'll write how this should work in general. At the top-most level the user passes a strategy (doesn't matter which one). This strategy at the top-most level can also be created based on the geometry if the user doesn't pass it explicitly. Then the strategy is passed below into the internals of an algorithm. This strategy defines the CS of all of the operations below (or defines all of the operations themselves as sub-strategies, it doesn't matter). What matters is that the information about the CS is passed by the user with the top-most level strategy.

So here `expand` is called. If the user's intention (by proxy) is to perform expansion of a box with some other geometry in a specific CS the following strategies should be used:
* cartesian
  * point - `strategy::expand::cartesian_point`
  * box - `strategy::expand::cartesian_box`
  * segment - `strategy::expand::cartesian_segment`
* spherical
  * point - `strategy::expand::spherical_point`
  * box - `strategy::expand::spherical_box`
  * segment - `strategy::expand::spherical_segment`
* geographic
  * point - `strategy::expand::spherical_point`
  * box - `strategy::expand::spherical_box`
  * segment - `strategy::expand::geographic_segment`

In this case we're talking about `expand(box, point)` so the strategy is either `strategy::expand::cartesian_point` or `strategy::expand::spherical_point`. And at the top-most level of `buffer` the strategy created is the `IntersectionStrategy`. So one of the two `expand` strategies has to be derived from `IntersectionStrategy`. At least this is how it is right now, if we change it in the future it might be different (e.g. it might be derived from an `UmbrellaStrategy`). But right now, the `IntersectionStrategy` directly only defines Box (`expand_box_strategy_type`) and Segment (`expand_strategy_type`) expand strategies. So you have several options here:
1. Take it from one of the underlying strategies, e.g.:
```
IntersectionStrategy::template point_in_geometry_strategy<BoxPoint, TurnPoint>::expand_point_strategy_type
```
2. Create default one based on `IntersectionStrategy::cs_tag`:
```
strategy::expand::services::default_strategy<point_tag, IntersectionStrategy::cs_tag>::type
```
This will probably not compile right now for spherical because for spherical `IntersectionStrategy` the `cs_tag` is `spherical_tag` and `default_strategy` is not specialized for this tag so with this you'd have to add specialization here:
https://github.com/boostorg/geometry/blob/develop/include/boost/geometry/strategies/spherical/expand_point.hpp#L205-L221
3. Create it based on `IntersectionStrategy::cs_tag` without using `default_strategy` with, e.g.: 
```
typedef typename boost::if_c
    <
        boost::is_same<typename IntersectionStrategy::cs_tag, cartesian_tag>::value,
        strategy::expand::cartesian_point,
        strategy::expand::spherical_point,
    >::type expand_point_strategy_type;

// you could add some check for safety or not:

static const bool cs_check =
    boost::is_same<typename IntersectionStrategy::cs_tag, cartesian_tag>::value
    || boost::is_same<typename IntersectionStrategy::cs_tag, spherical_tag>::value
    || boost::is_same<typename IntersectionStrategy::cs_tag, geographic_tag>::value;
BOOST_MPL_ASSERT(cs_check);
```

Reply:
Thank you.

I think you're missing one important point I make. **It does not compile.** And not because the CS is wrong, it does not compile because the expand strategy, defined in intersection strategy, is defined for `segment` and not on `box` or `point`.
Therefore **it does not compile if you expand with a point.**

And this is an important flaw of how the strategy is passed. We might solve it by adding two other strategies to intersection_strategy. But I'm really opposed to that idea.

I still don't know why expand strategy should be passed at all. For other strategies, I understand the rationale (distance, area, point-in-polygon, side, etc). But this one is just an implementation detail, different for different cs, true, but not user-dependent or earth-model dependent.